src/rtc_signaling.cpp

```diff
@@ -0,0 +1,430 @@
+
+			post(m_io_context, std::bind(&rtc_signaling::on_data_channel
+				, self
+				, error_code{}
+				, offer_id
+				, dc_
+			));
+		});
+
+		// We need to maintain the DataChannel alive
+		conn.data_channel = std::move(dc);
+	}
+}
+
+void rtc_signaling::process_offer(rtc_offer const& offer)
+{
+	if (auto it = m_connections.find(offer.id); it != m_connections.end()) {
+		// It seems the offer is from ourselves, ignore...
+		return;
+	}
+
+#ifndef TORRENT_DISABLE_LOGGING
+	debug_log("*** RTC signaling processing remote offer");
+#endif
+	auto& conn = create_connection(offer.id, [this, offer](error_code const& ec, std::string sdp) {
+		rtc_answer answer{offer.id, offer.pid, std::move(sdp)};
+		post(m_io_context, std::bind(&rtc_signaling::on_generated_answer
+			, this
+			, ec
+			, answer
+			, offer
+        ));
+	});
+
+	conn.pid = offer.pid;
+
+	try {
+		conn.peer_connection->setRemoteDescription({offer.sdp, "offer"});
+	}
+	catch(const std::exception &e) {
+#ifndef TORRENT_DISABLE_LOGGING
+		debug_log("*** Failed to set remote RTC offer: %s", e.what());
+#endif
+	}
+}
+
+void rtc_signaling::process_answer(rtc_answer const& answer)
+{
+	auto it = m_connections.find(answer.offer_id);
+	if (it == m_connections.end()) return;
+
+#ifndef TORRENT_DISABLE_LOGGING
+	debug_log("*** RTC signaling processing remote answer");
+#endif
+
+	connection& conn = it->second;
+	if (conn.pid)
+	{
+#ifndef TORRENT_DISABLE_LOGGING
+		debug_log("*** Local RTC offer already got an answer");
+#endif
+		return;
+	}
+
+	conn.pid = answer.pid;
+
+	try {
+		conn.peer_connection->setRemoteDescription({answer.sdp, "answer"});
+	}
+	catch(const std::exception &e) {
+#ifndef TORRENT_DISABLE_LOGGING
+		debug_log("*** Failed to set remote RTC answer: %s", e.what());
+#endif
+	}
+}
+
+rtc_signaling::connection& rtc_signaling::create_connection(rtc_offer_id const& offer_id, description_handler handler)
+{
+	if (auto it = m_connections.find(offer_id); it != m_connections.end())
+		return it->second;
+
+#ifndef TORRENT_DISABLE_LOGGING
+	debug_log("*** RTC signaling creating connection");
+#endif
+
+	rtc::Configuration config;
+	std::string stun_server = m_torrent->settings().get_str(settings_pack::webtorrent_stun_server);
+	if (!stun_server.empty())
+		config.iceServers.emplace_back(std::move(stun_server));
+
+	auto pc = std::make_shared<rtc::PeerConnection>(config);
+	auto weak_pc = make_weak_ptr(pc);
+	pc->onStateChange([this, weak_this = weak_from_this(), offer_id](rtc::PeerConnection::State state)
+	{
+		// Warning: this is called from another thread
+		auto self = weak_this.lock();
+        if (!self) return;
+
+		if (state == rtc::PeerConnection::State::Failed)
+		{
+			post(m_io_context, std::bind(&rtc_signaling::on_data_channel
+				, std::move(self)
+				, boost::asio::error::connection_refused
+				, offer_id
+				, nullptr
+			));
+		}
+    });
+
+	pc->onGatheringStateChange([this, weak_this = weak_from_this(), offer_id, handler, weak_pc](
```

Comment:
I prefer to capture a weak pointer to `this` and locking it when entering the callback rather than capturing `self` to avoid holding lasting backward strong references in peer connections.

Reply:
right, I'm suggesting *not* capturing `this`, but instead explicitly use the `shared_ptr` you get from locking the `weak_ptr`. It makes it a bit more explicit what's going on.

Reply:
Oh sorry I misread, sure :+1:

Reply:
to be clear, the capture expression would look like this:
```
[weak_this = weak_from_this(), offer_id, handler, weak_pc=make_weak_ptr(pc)]
```